### PR TITLE
#168188701 User signup endpoint

### PR DESCRIPTION
**Pull request by hustlaviola**
#### What does this PR do?

This PR implements the user signup API endpoint

#### Description of Tasks to be completed

- Setup test suite for user signup endpoint
- Install `bcrypt` to encrypt user password
- Install `cross-env` to set environment variables
- Install `babel-eslint` to lint all babel code
- Install `nodemon` to monitor changes in code
- Install `core-js` and `@babel/runtime-corejs3` to emulate a full ES2015+ environment
- Create a helper class to generate a token and encrypt the password
- Create a utility class to handle responses
- Create a service class to mediate between database and user controller
- Create a controller class to implement signup endpoint
- Add the route
- Test with postman
- Ensure all test cases are passing
- Document with Swagger

#### How should this be manually tested?

- Clone the Repo, change into the directory and run `npm install`
- Run `npm test` to see all test cases
- Run `npm dev` to start the server
- Make a `POST` request to `http://localhost:{PORT}/api/v1/users/signup` on Postman using valid credentials

Sample request body
```
{
    "email": "naimatdavid@mail.com",
    "firstname": "Ibrahim",
    "lastname": "Jibson",
    "password": "ChibuokemAlabi99"
}
```

#### Relevant pivotal tracker story

[168188701](https://www.pivotaltracker.com/story/show/168188701)

#### Screenshots

Screenshot (128)

### CODE REVIEW

**Review by Naimaoye — APPROVED** [review]

**Review by ibraheem-adeyemo — CHANGES_REQUESTED** [review]
Kindly change the user response object to {response status, token, user_id, email, user_role}. Endeavour to remove first name, last name and password Thaks.

**Comment by hustlaviola** [comment]
> Kindly change the user response object to {response status, token, user_id, email, user_role}. Endeavour to remove first name, last name and password Thaks.

Thank you, Ibrahim, feedback implemented. Roles and other fields will not be available when creating a new account. It will, however, be available on *Profile setup* API.

**Comment by hustlaviola** [comment]
> Kindly change the user response object to {response status, token, user_id, email, user_role}. Endeavour to remove first name, last name and password Thaks.

Thank you, Ibrahim, feedback implemented. However, roles and other fields will not be available when creating a new account. It will be available on *Profile setup* API.

**Review by ibraheem-adeyemo — APPROVED** [review]

**Comment by daviddamilola** [comment]
Great work, however, babel-polyfill is deprecated, we can use regenerator-runtime instead. https://babeljs.io/docs/en/next/babel-polyfill.html. 

The npm link is https://www.npmjs.com/package/regenerator-runtime
`npm install --save regenerator-runtime` then we can remove the `--require babel-polyfill` and use

``` cross-env NODE_ENV=test nyc mocha -r @babel/register -r regenerator-runtime ./src/tests --exit ```
in the test script.

Also, the test only checks for a user with valid details,  we can also write tests to check for other scenarios.

Finally, I propose the endpoint be changed to `/auth/signup`

**Comment by hustlaviola** [comment]
> Great work, however, babel-polyfill is deprecated, we can use regenerator-runtime instead. https://babeljs.io/docs/en/next/babel-polyfill.html.
> 
> The npm link is https://www.npmjs.com/package/regenerator-runtime
> `npm install --save regenerator-runtime` then we can remove the `--require babel-polyfill` and use
> 
> `cross-env NODE_ENV=test nyc mocha -r @babel/register -r regenerator-runtime ./src/tests --exit`
> in the test script.
> 
> Also, the test only checks for a user with valid details, we can also write tests to check for other scenarios.
> 
> Finally, I propose the endpoint be changed to `/auth/signup`

Feedback implemented. I didn't follow your suggestions about the `generator-runtime`, that's because I had already found a way around it before seeing your message. User validation is another story on PT and the PR will be raised shortly. The `/user/signup/` endpoint wasn't my choice either, it was already part of the codebase. I believe we can discuss it more. Please review the changes. Thanks, David.

File: src/controllers/UserController.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+   * @returns {object} JSON response
+   * @memberof UserController
+   */
+  static async signup(req, res) {
+    const {
+      firstname, lastname, email, password
+    } = req.body;
+    const user = {
+      firstname, lastname, email, password
+    };
```
Comment (ascii-dev) [inline src/controllers/UserController.js:27]: ```const user = {...req.body}```

File: .env.example [hunk 2]
```diff
@@ -1,3 +1,3 @@
-PORT=5000
+PORT=3000
```
Comment (ascii-dev) [inline .env.example]: Any reason why you changed this from 5000 to 3000?

File: src/database/migrations/20190903175944-create-user.js [hunk 3]
```diff
@@ -6,12 +6,12 @@ module.exports = {
...
+    firstName: { type: Sequelize.STRING, allowNull: false },
+    lastName: { type: Sequelize.STRING, allowNull: false },
+    email: { type: Sequelize.STRING, allowNull: false, unique: true },
     password: { type: Sequelize.STRING, allowNull: false },
     createdAt: { type: Sequelize.DATE, allowNull: false },
     updatedAt: { type: Sequelize.DATE, allowNull: false }
   }),
-  down: (queryInterface) => queryInterface.dropTable('Users')
+  down: queryInterface => queryInterface.dropTable('Users')
 };
```
Comment (ascii-dev) [inline src/database/migrations/20190903175944-create-user.js:17]: Why did you change this from `create-users` to `create-user`?

File: .env.example (hunk 2 above)
Comment (hustlaviola) [reply .env.example]: At a point, I felt it will be easier to tell a user to use the PORT in `./src/index.js` file when manually testing but that's not necessary

File: src/database/migrations/20190903175944-create-user.js (hunk 3 above)
Comment (hustlaviola) [reply src/database/migrations/20190903175944-create-user.js:17]: I created another table as suggested. But I used `user` instead of `users` since Sequelize will automatically pluralize it.

Comment (ascii-dev) [reply src/database/migrations/20190903175944-create-user.js:17]: Did you delete the current migration file before creating the new one? @hustlaviola

Comment (hustlaviola) [reply src/database/migrations/20190903175944-create-user.js:17]: Yes

**Review by ascii-dev — APPROVED** [review]
LGTM
